cmd/scriggo/run.go

```diff
@@ -1,84 +1,289 @@
+	if flags.s >= -1 {
+		asm := template.Disassemble(flags.s)
+		_, _ = os.Stderr.Write(asm)
+	}
+
+	out := os.Stdout
+	if flags.o != "" {
+		var fi *os.File
+		fi, err = openRenderOut(flags.o, path.Base(name))
+		if err != nil {
+			return err
+		}
+		defer func() {
+			err = fi.Close()
+		}()
+		out = fi
+	}
 
-	// Validate command line arguments.
-	if len(os.Args) != 2 {
-		_, _ = fmt.Fprintf(os.Stderr, "usage: %s program.go", os.Args[0])
-		os.Exit(1)
+	if flags.metrics {
+		start = time.Now()
 	}
-	file := os.Args[1]
-	name := filepath.Base(file)
-	if name == ".go" {
-		fmt.Printf("%s: invalid file name", name)
-		os.Exit(1)
+
+	// Run the template.
+	err = template.Run(out, nil, nil)
+
+	if flags.metrics {
+		runTime := time.Since(start)
+		_, _ = fmt.Fprintf(os.Stderr, "\n     %12s  %12s  %12s  %s\n", "Build", "Run", "Total", "File")
+		_, _ = fmt.Fprintf(os.Stderr, "     %12s  %12s  %12s  %s\n", "-----", "---", "-----", "----")
+		_, _ = fmt.Fprintf(os.Stderr, "     %12s  %12s  %12s  %s\n", buildTime, runTime, buildTime+runTime, name)
 	}
-	ext := filepath.Ext(name)
-	if ext != ".go" {
-		fmt.Printf("%s: extension must be \".go\"\n", name)
-		os.Exit(1)
+
+	return err
+}
+
+// parseFormat parses and returns a format.
+func parseFormat(s string) (scriggo.Format, error) {
+	switch s {
+	case "Text":
+		return scriggo.FormatText, nil
+	case "HTML":
+		return scriggo.FormatHTML, nil
+	case "Markdown":
+		return scriggo.FormatMarkdown, nil
+	case "CSS":
+		return scriggo.FormatCSS, nil
+	case "JS":
+		return scriggo.FormatJS, nil
+	case "JSON":
+		return scriggo.FormatJSON, nil
 	}
+	return 0, fmt.Errorf("invalid format %s, format can be Text, HTML, Markdown, CSS, JS or JSON", s)
+}
+
+// formatFS is a file system that implements scriggo.FormatFS.
+type formatFS struct {
+	fs.FS
+	format scriggo.Format
+}
 
-	// Read the source of the program.
-	src, err := os.ReadFile(file)
+func (fsys formatFS) Format(name string) (scriggo.Format, error) {
+	return fsys.format, nil
+}
+
+// openRenderOut opens and returns a file named name, creating or truncating
+// it. If it is a directory it opens and returns the file named base in
+// the directory.
+func openRenderOut(name, base string) (*os.File, error) {
+	fi, err := os.OpenFile(name, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0666)
 	if err != nil {
-		_, _ = fmt.Fprintln(os.Stderr, err)
-		os.Exit(2)
+		st, err2 := os.Stat(name)
+		if err2 != nil {
+			return nil, err2
+		}
+		if !st.IsDir() {
+			return nil, err
+		}
+		fi, err = os.OpenFile(path.Join(name, base), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0666)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return fi, nil
+}
+
+// parseConstants parses s as a sequence of name=value pairs where name is a
+// Go identifier (excluded the blank identifier and Go predefined identifiers)
+// and value is a string literal, a number literal, true or false. It adds
+// each name/value pair to globals.
+func parseConstants(s string, globals native.Declarations) error {
```

Comment:
Consider this template (file _index.html_):

```
{{ a }}{{ b }}
```

Running it using:

```
scriggo run -const a=10 -const b=20 index.html
```

results in this error:

```
index.html:1:4: undefined: a
```

Reply:
Do in commit "review: allow multiple 'const' flags".

Reply:
Did in commit "review: allow multiple 'const' flags".